Plot chat bugs

Just updated to v3.1.3 on Frostcast.

We've got some problems with plot chat.

First of all, it didn't appear to work at all, /plot chat said on and off messages, yet messages were still going to global. Found a setting in settings.yml named chat, enabled it, and it appeared to be working again at first.

Bug report from a player:
> ok so, if a player goes into p chat. Some times it will not allow them to chat unless they leave the plot. Other times they are stuck in just 1 plots chat and cant talk in global

```
2015-08-22 20:08:47 [ERROR] Could not pass event AsyncPlayerChatEvent to PlotSquared v3.1.3
org.bukkit.event.EventException
        at org.bukkit.plugin.java.JavaPluginLoader$1.execute(JavaPluginLoader.java:310) ~[spigot-1.8.8.jar:git-Spigot-d97e08b-880a532]
        at org.bukkit.plugin.RegisteredListener.callEvent(RegisteredListener.java:62) ~[spigot-1.8.8.jar:git-Spigot-d97e08b-880a532]
        at org.bukkit.plugin.SimplePluginManager.fireEvent(SimplePluginManager.java:502) [spigot-1.8.8.jar:git-Spigot-d97e08b-880a532]
        at org.bukkit.plugin.SimplePluginManager.callEvent(SimplePluginManager.java:484) [spigot-1.8.8.jar:git-Spigot-d97e08b-880a532]
        at net.minecraft.server.v1_8_R3.PlayerConnection.chat(PlayerConnection.java:1084) [spigot-1.8.8.jar:git-Spigot-d97e08b-880a532]
        at net.minecraft.server.v1_8_R3.PlayerConnection.a(PlayerConnection.java:1022) [spigot-1.8.8.jar:git-Spigot-d97e08b-880a532]
        at net.minecraft.server.v1_8_R3.PacketPlayInChat$1.run(PacketPlayInChat.java:39) [spigot-1.8.8.jar:git-Spigot-d97e08b-880a532]
        at java.util.concurrent.Executors$RunnableAdapter.call(Unknown Source) [?:1.7.0_17]
        at java.util.concurrent.FutureTask$Sync.innerRun(Unknown Source) [?:1.7.0_17]
        at java.util.concurrent.FutureTask.run(Unknown Source) [?:1.7.0_17]
        at java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source) [?:1.7.0_17]
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source) [?:1.7.0_17]
        at java.lang.Thread.run(Unknown Source) [?:1.7.0_17]
Caused by: java.lang.StringIndexOutOfBoundsException: String index out of range: 4
        at java.lang.String.charAt(Unknown Source) ~[?:1.7.0_17]
        at java.util.regex.Matcher.appendReplacement(Unknown Source) ~[?:1.7.0_17]
        at java.util.regex.Matcher.replaceAll(Unknown Source) ~[?:1.7.0_17]
        at java.lang.String.replaceAll(Unknown Source) ~[?:1.7.0_17]
        at com.plotsquared.bukkit.listeners.PlayerEvents.onChat(PlayerEvents.java:634) ~[?:?]
        at sun.reflect.GeneratedMethodAccessor205.invoke(Unknown Source) ~[?:?]
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(Unknown Source) ~[?:1.7.0_17]
        at java.lang.reflect.Method.invoke(Unknown Source) ~[?:1.7.0_17]
        at org.bukkit.plugin.java.JavaPluginLoader$1.execute(JavaPluginLoader.java:306) ~[spigot-1.8.8.jar:git-Spigot-d97e08b-880a532]
        ... 12 more
```

--- a/src/main/java/com/plotsquared/bukkit/listeners/PlayerEvents.java
+++ b/src/main/java/com/plotsquared/bukkit/listeners/PlayerEvents.java
@@ -610,10 +610,17 @@ public class PlayerEvents extends com.plotsquared.listener.PlotListener implemen
             return;
         }
         final PlotWorld plotworld = PS.get().getPlotWorld(world);
-        final PlotPlayer plr = BukkitUtil.getPlayer(player);
-        if (!plotworld.PLOT_CHAT && (plr.getMeta("chat") == null || !(Boolean) plr.getMeta("chat"))) {
+
+        if (!plotworld.PLOT_CHAT) {
             return;
         }
+
+        final PlotPlayer plr = BukkitUtil.getPlayer(player);
+
+        if (!plr.getAttribute("chat")) {
+            return;
+        }
+
         final Location loc = BukkitUtil.getLocation(player);
         final Plot plot = MainUtil.getPlot(loc);
         if (plot == null) {
@@ -631,9 +638,10 @@ public class PlayerEvents extends com.plotsquared.listener.PlotListener implemen
                 recipients.add(p);
             }
         }
-        format = format.replaceAll("%plot_id%", id.x + ";" + id.y).replaceAll("%sender%", sender).replaceAll("%msg%", message);
+        format = format.replaceAll("%plot_id%", id.x + ";" + id.y).replaceAll("%sender%", "%s").replaceAll("%msg%", "%s");
         format = ChatColor.translateAlternateColorCodes('&', format);
         event.setFormat(format);
+        event.setMessage(message);
     }
 
     @EventHandler(priority = EventPriority.HIGHEST)

--- a/src/main/java/com/plotsquared/sponge/object/SpongePlayer.java
+++ b/src/main/java/com/plotsquared/sponge/object/SpongePlayer.java
@@ -4,9 +4,14 @@ import java.util.Date;
 import java.util.HashSet;
 import java.util.UUID;
 
+import org.spongepowered.api.data.key.Keys;
+import org.spongepowered.api.data.manipulator.mutable.TargetedLocationData;
+import org.spongepowered.api.data.value.mutable.Value;
 import org.spongepowered.api.entity.player.Player;
 import org.spongepowered.api.entity.player.gamemode.GameMode;
 import org.spongepowered.api.entity.player.gamemode.GameModes;
+import org.spongepowered.api.text.Texts;
+import org.spongepowered.api.text.chat.ChatType;
 import org.spongepowered.api.text.chat.ChatTypes;
 
 import com.flowpowered.math.vector.Vector3d;
@@ -100,7 +105,7 @@ public class SpongePlayer extends PlotPlayer {
 
     @Override
     public void sendMessage(String message) {
-        player.sendMessage(ChatTypes.CHAT, message);
+        player.sendMessage(ChatTypes.CHAT, Texts.of(message));
     }
 
     @Override

